modules/core/src/lib/attribute.js

```diff
@@ -184,29 +184,35 @@ export default class Attribute extends BaseAttribute {
     return false;
   }
 
-  updateBuffer({numInstances, bufferLayout, data, props, context}) {
+  updateBuffer({numInstances, bufferLayout, data, startIndex, endIndex, props, context}) {
     if (!this.needsUpdate()) {
       return false;
     }
 
     const state = this.userData;
 
-    const {update, noAlloc} = state;
+    const {update} = state;
 
     let updated = true;
     if (update) {
       // Custom updater - typically for non-instanced layers
-      update.call(context, this, {data, props, numInstances, bufferLayout});
-      if (noAlloc || this.constant || !this.buffer) {
+      update.call(context, this, {data, startIndex, endIndex, props, numInstances, bufferLayout});
+      if (this.constant || !this.buffer || this.buffer.byteLength < this.value.byteLength) {
         // Full update
         this.update({
           value: this.value,
           constant: this.constant
         });
       } else {
+        const startOffset = Number.isFinite(startIndex) ? this._getVertexOffset(startIndex) : 0;
+        const endOffset = Number.isFinite(endIndex)
+          ? this._getVertexOffset(endIndex)
+          : numInstances * this.size;
```

Comment:
Does this only support instanced data? This wouldn't work for a custom geometry layer?